src/main/ruby/core/io.rb

```diff
@@ -146,13 +146,30 @@ def unshift(str, start_pos)
     end
 
     def fill(io)
-      Truffle.primitive :iobuffer_fill
-
       unless io.kind_of? IO
-        return fill(io.to_io)
+        io = io.to_io
       end
 
-      raise PrimitiveFailure, 'InternalBuffer#fill primitive failed'
+      max = 8192
+      left = @total - @used
+      count = left < max ? left : max
+
+      begin
+        buffer = Truffle::POSIX.read_string(io.descriptor, count)
+      rescue Errno::EAGAIN
+        retry
+      end
+      bytes_read = buffer ? buffer.bytesize : 0
+      if bytes_read > 0
+        # Detect if another thread has updated the buffer
+        # and now there isn't enough room for this data.
+        if bytes_read > (@total - @used)
```

Comment:
This is still racy, isn't it? Two threads could both fail this check and then both try to fill the buffer causing one to overflow.

Reply:
Looking at it more, it was ported from Java. But the question still stands.

Reply:
Of course. But `fill` seems to be called only by `fill_from` which takes a monitor on the buffer.
So it should be fine if other operations touching the buffer also take the monitor (it seems unshift does not).
But this is out of scope of this PR.